add `--gradio-inpaint-tool` and option to specify `color-sketch`

kavorite opened the pull request:
Following a [comment] by @matrix4767 on the discussion post for [img2img-color-sketch], I implemented rudimentary functionality to let end users color on top of their uploaded images in the browser to better guide the inpainting process. The new feature infers a mask for diffusion sampling by caching the image upon its first upload. It then cross-references the cache with the subsequent edits of the image to determine altered regions after the end user is allowed to paint on top.

The state machine that implements this functionality is a little wonky, and I didn't want to surprise consumers of this repository who are already used to the other way of doing things. So, following in line with [img2img-color-sketch], new behavior can be enabled by appending `--gradio-inpaint-tool color-sketch` to the command line arguments, with the prior behavior, `sketch`, set as the default.

## Caveat Emptor
I couldn't figure out how to implement the state machine such that it ignores when an image is merely _edited,_ as opposed to cleared and reuploaded. GradIO's [docs] indicate that the `Image` component has `upload` and `change` events, which might lead one to believe these hooks have distinct triggers, but they seem to fire when images are edited, as well. I could perhaps have wired up the `edit` event to _cancel_ the `upload` event using coroutines if it turned out to be more exclusive, but that would have been more complicated, might have required enabling [queue] functionality, and could have introduced a race condition in cases where the events did not fire in the expected order.

Instead of heading further into the weeds, I opted to check whether channel vectors at corresponding locations between the two images are exactly equal, given that the two images are the same size. So, one possible issue is that a fresh upload might contain a pixel-perfect color match on top of the same spatial coordinate, and so falsely be identified as an edit. However, there's a low probability of this happening in different images given uniformly distributed values for each pixel's channels-- of which there are four, in this case $(R, G, B, A)$. Supposing we drop Alpha (maybe the end user uploaded a JPEG), we still only have 

$$
P(\mathrm{collision}) = \frac{1}{256\mathrm{\ possible\ values\ }^{3\ \mathrm{channels}}} \approx 5 \times 10^{-8}
$$

or, a false positive on the order of once every hundred million fresh uploads.

I recognize there are flaws in assuming conditional independence between regions of spatially correlated pixels or between the color channels, but pragmatically, in order for this heuristic to fail, the end user would have to paint over the image _completely_ or upload a new image of an identical size, with an exactly identical pixel located at an exactly identical spatial coordinate. I found a helpful [blog] with some alternative computer vision algorithms designed for detecting duplicates that I'll consider implementing if branching on the chosen condition is found to be insufficient.

[queue]: https://gradio.app/docs/#queue
[docs]: https://gradio.app/docs/#image
[img2img-color-sketch]: https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/2329
[comment]: https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/2329#discussioncomment-3942624
[blog]: https://www.hackerfactor.com/blog/?/archives/529-Kind-of-Like-That.html

kavorite committed:
add gradio-inpaint-tool; color-sketch

152334H commented:
This is a surprise. In #3865 I expected it would've been too difficult to use the gradio color-sketch tool for this purpose. Glad to see someone has attempted it anyway!

I'm not sure I understand your code correctly, but it seems like the paint strokes are not affected by the mask blur (although the generated mask is). Could you post some examples from your implementation?

kavorite commented:
> This is a surprise. In #3865 I expected it would've been too difficult to use the gradio color-sketch tool for this purpose. Glad to see someone has attempted it anyway!
> 
> I'm not sure I understand your code correctly, but it seems like the paint strokes are not affected by the mask blur (although the generated mask is). Could you post some examples from your implementation?

Certainly! Here's a little experiment I tossed in the UI:

![Annotated photograph of an apple](https://user-images.githubusercontent.com/15897272/200503960-946f00b1-4822-4570-a6c6-120eed552113.png)

And the result, inpainted with SDv1-4--
> apple with a bite taken out of it
Steps: 20, Sampler: Euler a, CFG scale: 7, Seed: 977361381, Size: 512x512, Model hash: 7460a6fa, Denoising strength: 0.75, Clip skip: 2, Mask blur: 4
![Inpainted photograph of an apple](https://user-images.githubusercontent.com/15897272/200503820-36bc9a9b-36ce-410d-8d03-aae2fd05d41d.png)

Re: 
> it seems like the paint strokes are not affected by the mask blur (although the generated mask is).

I actually haven't tried modifying this implementation to do that. Did it give better results for you? I wasn't aware of any semantic editing methods that blurred the paint applied for guidance purposes, so I'm unsure of how to approach it, and I only just started actually grepping around this codebase a matter of days ago. Tracing it in the debugger seems to be rather slow, especially on my system, so I'd greatly appreciate some pointers. Perhaps I can try that in a local branch and see how it performs.

152334H commented:
I'm still checking things out, but as a short fyi, you might've forgotten to commit the additional `init_img_with_mask_orig` argument in the call to `img2img` 

(around `modules/ui.py:944`)

152334H commented:
I'm having problems with a test of this on my end.

I'm using this image:

![istockphoto-475190475-612x612](https://user-images.githubusercontent.com/54623771/200511818-f30e00a0-09df-473a-a380-f90c9d671335.jpg)

Inpainted this way:

![IMG_20221108_162041](https://user-images.githubusercontent.com/54623771/200511992-c37393d7-109a-4a55-9794-eada794aa659.jpg)

And I get these results:
![IMG_20221108_162222](https://user-images.githubusercontent.com/54623771/200512169-b2d71756-b922-437c-ba7b-391ae7396da7.jpg)

I feel like there should be a bite at the left side of these images, but I would not be surprised if I was using the tool wrong somehow. Are there problems with what I'm attempting?

kavorite committed:
add new color-sketch state to img2img invocation

kavorite commented:
Ah. Perhaps the model sees the blur in the mask and tries to fill in the blurred content (it probably sees a pinkish splotch there). You might try setting the mask blur lower for your test? I also see that you have a different model hash, although I'm not sure whether I should expect that to matter too much, since I only used a CompVis checkpoint.

kavorite committed:
blur mask with color-sketch + add paint transparency slider

kavorite commented:
@152334H wanted to let you know that I implemented your suggestion and thank you for your helpful comment-- certainly works a lot better to have the blur, the model resolves the resulting ambiguities by re-establishing the continuity of partially obscured objects, producing images that are less cluttered with noise and artifacts. I also implemented a transparency slider so that the end-user has nominal control over the "weight" of their mask during the compositing step. 

For such a simple change this was surprisingly effective. Makes me wonder if there are any other denoising procedures that could be helpful: Blur has its own side effects that are sometimes a bit unhelpful, particularly when trying to perform inpainting of overlapping subjects, where objects can become fused, or on flat-colored illustrations, where the models' tendencies to dream inappropriate lights and shadows into existence can appear particularly jarring and out of place.

152334H commented:
Both changes are really cool, and working on my end. I have nothing more to suggest; it looks great.

ShadeRedeye commented:
usually using this color sketch,but there a problem with "undo" that cause previous painting still intact to a new image when dragging a image without delete (or click x button) a previous one

also could it possible to add highres fix to a inpaint/img2img tab

152334H commented:
> also could it possible to add highres fix to a inpaint/img2img tab

"High res fix" without scale latent enabled is functionally identical to img2img on higher resolutions.

willhsmit commented:
Tried out this tool and it is working well for me.

When this flag is on, is there a way to mask areas without overpainting them? For example, if I want to inpaint a larger area, but only want to explicitly provide my own color 'instruction' for a specific area and let the inpainting use the original image as a starting point for the rest of the area (perhaps to let in make subtler adjustments there for a cleaner boundary merge).

skdursh commented:
Hi, I'm a bit new to Github and am just wondering how I can try this out since it isn't yet part of the main repository. Do I just use Git Bash with the checkout command pointed at this specific pull request or do I need to clone your fork and completely redownload all the requirements to use this? Thanks for any help someone might give me!

152334H commented:
> Hi, I'm a bit new to using Github via the actual Git protocol and am just wondering how I can try this out since it isn't yet part of the main repository? Do I just use Git Bash with the checkout command pointed at this specific pull request or do I need to clone your fork and completely redownload all the requirements to use this? Thanks for any help someone might give me!

Start by adding the forked repository as a "remote":
```sh
git remote add kavorite https://github.com/kavorite/stable-diffusion-webui
git fetch --all # this "downloads" the remote
```
Switch to the forked repository
```sh
git checkout -b inpainting kavorite/color-sketch-inpainting
# if you get an error at this point, try running `git stash`
```
Because the fork (kavorite's repository) is much older (many commits behind) than the main repository (AUTOMATIC's), you will need to "rebase" the fork to update it:
```sh
git rebase origin/master
# If you ran `git stash` previously, you probably want to run `git stash pop` now
```

adamz5k commented:
Works decently well for me.

I'd like to see at least one of two things (preferably both) before I'd consider using it:
1) A slider that expands the mask, rather than just blurs it, to be able to apply the mask to a set distance outside the painted color, with full weight.
2) The ability to toggle between painting color and painting a mask, as suggested by @skdursh

I also got the same bug as @ShadeRedeye

matrix4767 commented:
Possible to turn this into an extension?

lbeltrame commented:
FTR, this breaks inpainting.

lbeltrame commented on modules/ui.py:
```diff
@@ -840,11 +840,22 @@ def create_ui(wrap_gradio_gpu_call):
 
+                        def update_orig(image, state):
+                            if image is not None:
+                                same_size = state is not None and state.size == image.size
```
This breaks because `state` is a dict and doesn't have that attribute.

Jonseed commented:
I like the ability to paint in inpainting mode, but I also want to be able to mask as in prior behavior. Is there a toggle?

willhsmit commented:
> I like the ability to paint in inpainting mode, but I also want to be able to mask as in prior behavior. Is there a toggle?

One workaround I found is starting two copies of webui, one with the flag and one without. They start on different ports when running locally, so you can inpaint in color sketch mode and then move the image over to a non-colorsketch tab to use a slightly larger mask to clean up the boundaries, without having to restart the app.

zeoo01 commented:
I'm pretty new to this and I installed stable diffusion manually by downloading the zip, is there a way to implement this with some command or do I have to download something or how is it?

Jonseed commented:
> I'm pretty new to this and I installed stable diffusion manually by downloading the zip, is there a way to implement this with some command or do I have to download something or how is it?

You add "--gradio-inpaint-tool color-sketch" to the command line arguments in webui-user.bat

zeoo01 commented:
> > I'm pretty new to this and I installed stable diffusion manually by downloading the zip, is there a way to implement this with some command or do I have to download something or how is it?
> 
> You add "--gradio-inpaint-tool color-sketch" to the command line arguments in webui-user.bat

didnt work :(
